service/s3/unmarshal.go

```diff
@@ -0,0 +1,120 @@
+package s3
+
+import (
+	"bytes"
+	"encoding/xml"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	restlegacy "github.com/aws/aws-sdk-go-v2/private/protocol/rest"
+)
+
+/*
+   GetObject operation doesn't support 200 status code with errors.
+
+*/
+
+// protoGetObjectUnmarshaler defines unmarshaler for ProtoGetObject operation
+type protoGetObjectUnmarshaler struct {
+	output *GetObjectOutput
+}
+
+// unmarshalOperation is the top level method used with a handler stack to unmarshal an operation response
+// This method calls appropriate unmarshal shape functions as per the output shape and protocol used by the
+// service.
+func (u protoGetObjectUnmarshaler) unmarshalOperation(r *aws.Request) {
+
+	var (
+		buff       bytes.Buffer
+		body       io.Reader
+		decoder    *xml.Decoder
+		startToken xml.Token
+		err        error
+	)
+
+	if r.HTTPResponse.Body != nil {
+		body = io.TeeReader(r.HTTPResponse.Body, &buff)
+		decoder = xml.NewDecoder(body)
+		startToken, err = decoder.Token()
+		// we will encounter io.EOF if body does not have an xml
+		if err != nil && err != io.EOF {
+			r.Error = awserr.New(aws.ErrCodeSerialization, "Failed to decode response body with invalid XML", err)
+			return
+		}
```

Comment:
Don't want to attempt to parse the response body as XML unless the `isRequestError` returns true. Attempting to parse the body too soon like here will incorrectly result in fail to unmarshal XML errors.


Creating the decoder and getting first token should be moved within the `isREquestError` check below. This will also remove the need for optional nil pointers accessible at the top level of this method.